LOOP-1392 Alerts should be stored in iOS persistent storage for a limited amount of time

darinkrauss opened the pull request:
- https://tidepool.atlassian.net/browse/LOOP-1392
- Add expireAfter and expireDate to AlertStore
- Purge expired alerts
- Add Core Data bulk add support to AlertStore
- Add simulated Core Data support to AlertStore

Requesting @rickpasetto and one other reviewer, please.

Commit:
LOOP-1392 Alerts should be stored in iOS persistent storage for a limited amount of time

- https://tidepool.atlassian.net/browse/LOOP-1392
- Add expireAfter and expireDate to AlertStore
- Purge expired alerts
- Add Core Data bulk add support to AlertStore
- Add simulated Core Data support to AlertStore

rickpasetto commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -119,6 +125,29 @@ extension AlertStore {
+// MARK: Alert Purging
+
+extension AlertStore {
+    var expireDate: Date {
```
[nit] could be private?

darinkrauss replied:
Pete requested that the "Simulated Core Data" code be separated from the rest of the code since it is primarily debug/test code. To accomplish that I pushed all related code for all stores into separate files named "<store>+SimulatedCoreData.swift" and added the code in an extension. Thus, this property must at least be internal.

rickpasetto commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -56,6 +60,7 @@ extension AlertStore {
             do {
                 try self.managedObjectContext.save()
                 self.log.default("Recorded alert: %{public}@", alert.identifier.value)
+                self.purgeExpired()
```
[nit] This, of course, means that if the user doesn't get an alert for a while, that the purging would not happen, possibly well beyond the 90 days.  I presume, in practice, this is not an issue?  (Not blocking, just wondering what the pattern normally is) Otherwise, we might want to `public`-ize `purgeExpired` to be run as part of a regular CGM update or (or some such).  Just a thought...

darinkrauss replied:
Yes, that is correct. I don't see that as an issue. If nothing has changed, then I see no reason to purge the cache. There will be, in another ticket, code that will purge all caches to shorter and shorter expirations upon device low storage conditions.

Besides, the total storage space for the alert store, even given a fairly large number of alerts/day, is zilch compared to the rest of the stores. That is, 90 days worth of simulated data for all of the other stores is 2GB, while this is maybe a few MB. And, if they really are getting lots of alerts every day, then the store will be purged frequently as well.

rickpasetto reviewed (approved):
Mostly nits, but other than that LGTM.

nhamming reviewed (approved):
LGTM

Commit:
Merge branch 'dev' into darinkrauss/LOOP-1392/expire-alerts

Commit:
Merge branch 'dev' into darinkrauss/LOOP-1392/expire-alerts

# Conflicts:
#	Loop/Managers/Alerts/AlertManager.swift
#	Loop/Managers/Alerts/AlertStore.swift

Commit:
Fix issue introduced during merge

Commit:
Merge branch 'dev' into darinkrauss/LOOP-1392/expire-alerts

Commit:
Use deleteObjects rather than purgeObjects for AlertStore

- Update tests to ensure alerts are not purged while testing

Commit:
Fix simulated core data for stored alerts to set issued and acknowledged dates

darinkrauss commented:
@rickpasetto Would you take a look at the changes since you last reviewed? I changed how StoredAlerts are purged, updated simulated alerts, and fixed some tests that break when alerts are purged during the tests.

rickpasetto commented on Loop/Managers/Alerts/AlertStore.swift:
```diff
@@ -291,7 +293,12 @@ extension Result where Success == Void {
 
 extension AlertStore {
-    public func addAlerts(alerts: [Alert]) -> Error? {
+    public struct DatedAlert {
```
[nit] If these are Test only, can you make them `internal` instead of `public`?

rickpasetto reviewed (approved):
One [nit] comment/request, other than that LGTM.

Commit:
Updates from PR review

Commit:
Merge branch 'dev' into darinkrauss/LOOP-1392/expire-alerts
